File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/PriceBulldozer.java
package cz.fb.manaus.reactor.price;

import com.google.common.base.Preconditions;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.PriceComparator;
import cz.fb.manaus.core.model.TradedVolume;
import cz.fb.manaus.reactor.rounding.RoundingService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.LinkedList;
import java.util.List;

@Component
public class PriceBulldozer {

    @Autowired
    private RoundingService roundingService;

    public List<Price> bulldoze(double threshold, List<Price> prices) {
        var sum = 0d;
        Preconditions.checkState(PriceComparator.ORDERING.isStrictlyOrdered(prices));
        var convicts = new LinkedList<Price>();
        var untouched = new LinkedList<Price>();
        for (var price : prices) {
            if (sum >= threshold) {
                untouched.add(price);
            } else {
                convicts.add(price);
            }
            sum += price.getAmount();
        }
        var priceMean = TradedVolume.getWeightedMean(convicts);
        if (priceMean.isPresent()) {
            var amount = convicts.stream().mapToDouble(Price::getAmount).sum();
            var price = roundingService.roundBet(priceMean.getAsDouble()).getAsDouble();
            untouched.addFirst(new Price(price, amount, prices.get(0).getSide()));
        }
        return untouched;
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/PriceBulldozer.kt
package cz.fb.manaus.reactor.price

import com.google.common.collect.Comparators
import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.PriceComparator
import cz.fb.manaus.core.model.getWeightedMean
import org.springframework.stereotype.Component


fun getWeightedMean(prices: List<Price>): Double? {
    return getWeightedMean(prices, Price::price, Price::amount)
}

@Component
object PriceBulldozer {

    fun bulldoze(threshold: Double, prices: List<Price>): List<Price> {
        var sum = 0.0
        check(Comparators.isInStrictOrder(prices, PriceComparator))
        val convicts = mutableListOf<Price>()
        val untouched = mutableListOf<Price>()
        for (price in prices) {
            if (sum >= threshold) {
                untouched.add(price)
            } else {
                convicts.add(price)
            }
            sum += price.amount
        }
        val priceMean = getWeightedMean(convicts)
        if (priceMean != null) {
            val amount = convicts.sumOf { it.amount }
            val price = Price.round(priceMean)
            untouched.add(0, Price(price, amount, prices[0].side))
        }
        return untouched
    }

}
